Email verification action code settings

onimisi opened the pull request:
## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation  changes
- [ ] Other... Please describe:

<!-- Fill out the relevant sections below and delete irrelevant sections. -->

## PR Checklist
Please check your PR fulfills the following requirements:

Bugfixes:
- [ ] Regression testing has been carried out using the [example project](https://github.com/dpa99c/cordova-plugin-firebasex-test) to ensure the intended bug is fixed and no regression bugs have been inadvertently introduced.

New features/enhancements:
- [ ] Exhaustive testing has been carried out for the new functionality
- [ ] Regression testing has been carried out to ensure no existing functionality is adversely affected
- [ ] Documentation has been added / updated
- [ ] The [example project](https://github.com/dpa99c/cordova-plugin-firebasex-test) has been update to validate/demonstrate the new functionality.

## What is the purpose of this PR?
<!-- Describe any current behavior that you are modifying, or link to a relevant issue. -->
<!-- Describe the new behaviour added/modified and its purpose. -->

## Does this PR introduce a breaking change?
- [ ] Yes
- [ ] No

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing plugin versions. -->

## What testing has been done on the changes in the PR?
<!-- e.g. if an example project exists for this plugin, has it been updated to test the new functionality? -->

## What testing has been done on existing functionality?
<!-- e.g. if an example project exists for this plugin, has been it been tested to ensure no regression bugs have been introduced? -->

## Other information

Commit:
add settings parameters to types/index

Commit:
add actionCodeSettings as a command argument

Commit:
use parameters for the actionCodeSettings values in both ios and android

Commit:
make actionCodeSetting input optional

Commit:
add types to the argument parameters for iOS, revert optional actionCodeSettings

Commit:
update readme.md

Commit:
make sure ActionCodeSettings is optional and doesn't break if it doesn't exist

Commit:
add optional parameters within the actionCodeSettings object

tzappia commented on src/android/FirebasePlugin.java:
```diff
@@ -1280,13 +1281,30 @@ public void sendUserEmailVerification(final CallbackContext callbackContext, fin
+                            actionCodeSettingsParams.getBoolean("installIfNotAvailable"),
+                            actionCodeSettingsParams.getString("minimumVersion"))
+                        .build();
                     if(user == null){
```
This user check is a basic validation that should happen before you do the argument processing.

tzappia commented on src/android/FirebasePlugin.java:
```diff
@@ -1280,13 +1281,30 @@ public void sendUserEmailVerification(final CallbackContext callbackContext, fin
                     FirebaseUser user = FirebaseAuth.getInstance().getCurrentUser();
+                    ActionCodeSettings actionCodeSettings = null;
+
+                    if(!args.isNull(0)) {
```
Declare / init the actionCodeSettings inside this block. You should only have to check for the existence once.

onimisi replied:
@tzappia I had to define it outside the block because the check for actionCodeSettings below didn't know about the variable since it's defined in an if block.

tzappia replied:
How about:
```
if(!args.isNull(0)) {
  JSONObject actionCodeSettingsParams = args.getJSONObject(0);
  ActionCodeSettings actionCodeSettings = ActionCodeSettings.newBuilder()...
  handleTaskOutcome(user.sendEmailVerification(actionCodeSettings), callbackContext);
} else {
  handleTaskOutcome(user.sendEmailVerification(), callbackContext);
}
```

tzappia commented on src/android/FirebasePlugin.java:
```diff
@@ -1280,13 +1281,30 @@ public void sendUserEmailVerification(final CallbackContext callbackContext, fin
+
+                    if(!args.isNull(0)) {
+                      JSONObject actionCodeSettingsParams = args.getJSONObject(0);
+                      actionCodeSettings = ActionCodeSettings.newBuilder()
```
If any of these params are mandatory, check them and return an error if they are not present. Unless the builder will throw a specific error?

onimisi replied:
url is the only required parameter. The builder does throw its own error when the url is missing

tzappia commented on src/ios/FirebasePlugin.m:
```diff
@@ -889,18 +889,52 @@ - (void)updateUserEmail:(CDVInvokedUrlCommand*)command {
+                                           minimumVersion:[NSString stringWithString:[actionCodeSettingsParams objectForKey:@"minimumVersion"]]];
+            }
+        }
         if(!user){
```
Similar comments to the Java flow. Check the user first. Only check / init actionCodeSettings once.

Commit:
perform check for if user exists before processing the action code parameters.
make accessing the values of arguments optional to prevent nil error messages.
make only the url parameter required.

tzappia commented on src/android/FirebasePlugin.java:
```diff
@@ -1280,13 +1281,29 @@ public void sendUserEmailVerification(final CallbackContext callbackContext, fin
+                        handleTaskOutcome(user.sendEmailVerification(), callbackContext);
+                    }
+
+                    } catch (Exception e) {
```
Outdent this line to match where it was originally.

Commit:
line formatting

onimisi committed:
Update README.md

add spaces back
